File: robocode/robocode/render/RenderObject.java
/*******************************************************************************
 * Copyright (c) 2001, 2007 Mathew A. Nelson and Robocode contributors
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Common Public License v1.0
 * which accompanies this distribution, and is available at
 * http://robocode.sourceforge.net/license/cpl-v10.html
 *
 * Contributors:
 *     Flemming N. Larsen
 *     - Initial implementation
 *******************************************************************************/
package robocode.render;


import java.awt.Graphics2D;
import java.awt.Rectangle;
import java.awt.geom.AffineTransform;


/**
 * The base of all renderable objects.
 * An <code>RenderObject</code> is an object that can be painted and transformed.
 *
 * @author Flemming N. Larsen (original)
 */
public class RenderObject {

	/** Base transform, e.g. the initial rotation and translation */
	protected AffineTransform baseTransform;

	/** Current transform that is concatenated with the base transform */
	protected AffineTransform transform;

	/** Current frame that must be rendered */
	protected int frame;

	/**
	 * Constructs a new <code>RenderObject</code>.
	 */
	public RenderObject() {
		baseTransform = new AffineTransform();
		transform = new AffineTransform();
	}

	/**
	 * Constructs a new <code>RenderObject</code> that is a copy of another
	 * <code>RenderObject</code>.
	 */
	public RenderObject(RenderObject ro) {
		baseTransform = new AffineTransform(ro.baseTransform);
		transform = new AffineTransform(ro.transform);
		frame = ro.frame;
	}

	/**
	 * Sets the base transform which is pre-concatenated with the current transform.
	 *
	 * @param Tx the new transform
	 */
	public void setBaseTransform(AffineTransform Tx) {
		baseTransform.setTransform(Tx);
	}

	/**
	 * Returns a copy of the base transform
	 *
	 * @return a copy of the base transform
	 */
	public AffineTransform getBaseTransform() {
		return new AffineTransform(baseTransform);
	}

	/**
	 * Sets the current transform, which is concatenated with the base transform.
	 *
	 * @param Tx the new transform
	 */
	public void setTransform(AffineTransform Tx) {
		transform.setTransform(Tx);
		transform.concatenate(baseTransform);
	}

	/**
	 * Returns a copy of the current transform
	 *
	 * @return a copy of the current transform
	 */
	public AffineTransform getTransform() {
		return transform;
	}

	/**
	 * Sets the current frame number
	 *
	 * @param frame the new frame number
	 */
	public void setFrame(int frame) {
		this.frame = frame;
	}

	/**
	 * Returns the current frame number
	 *
	 * @return the current frame number
	 */
	public int getFrame() {
		return frame;
	}

	/**
	 * Paint this object.
	 * This method must be overridden in derived classes.
	 *
	 * @param g the graphics context where this object must be painted.
	 */
	public void paint(Graphics2D g) {}

	/**
	 * Returns the bounds of this object based on it's current transform.
	 * This method must be overridden in derived classes.
	 *
	 * @return the bounds of this object based on it's current transform.
	 */
	public Rectangle getBounds() {
		return new Rectangle();
	}

	/**
	 * Returns a copy of this object.
	 * This method must be overridden in derived classes.
	 *
	 * @return a copy of this object.
	 */
	public RenderObject copy() {
		return new RenderObject(this);
	}
}

File: nrobocodeui/nrobocodeui/gfx/RenderObject.cs
// ****************************************************************************
// Copyright (c) 2001, 2007 Mathew A. Nelson and Robocode contributors
// All rights reserved. This program and the accompanying materials
// are made available under the terms of the Common Public License v1.0
// which accompanies this distribution, and is available at
// http://robocode.sourceforge.net/license/cpl-v10.html
// 
// Contributors:
// Flemming N. Larsen
// - Initial implementation
// Pavel Savara
// - Conversion
// *****************************************************************************

using System;
using System.Drawing;
using System.Drawing.Drawing2D;

namespace nrobocodeui.gfx
{
    /// <summary> The base of all renderable objects.
    /// An <code>RenderObject</code> is an object that can be painted and transformed.
    /// </summary>
    /// <author>
    /// Flemming N. Larsen (original)
    /// </author>
    public abstract class RenderObject : ICloneable
    {
        /// <summary> 
        /// Sets the base transform which is pre-concatenated with the current transform.
        /// Returns a copy of the base transform
        /// </summary>
        public virtual Matrix BaseTransform
        {
            get { return baseTransform.Clone(); }
            set { baseTransform = value.Clone(); }
        }

        /// <summary> 
        /// Sets the current transform, which is concatenated with the base transform.
        /// Returns a copy of the current transform
        /// </summary>
        public virtual Matrix Transform
        {
            get { return transform; }

            set
            {
                transform = value.Clone();
                transform.Multiply(baseTransform, MatrixOrder.Append);
            }
        }

        /// <summary> 
        /// Sets the current frame number
        /// Returns the current frame number
        /// </summary>
        public virtual int Frame
        {
            get { return frame; }

            set { frame = value; }
        }

        /// <summary> 
        /// Returns the bounds of this object based on it's current transform.
        /// </summary>
        public abstract Rectangle Bounds { get; }

        /// <summary>Base transform, e.g. the initial rotation and translation </summary>
        protected internal Matrix baseTransform;

        /// <summary>Current transform that is concatenated with the base transform </summary>
        protected internal Matrix transform;

        /// <summary>Current frame that must be rendered </summary>
        protected internal int frame;

        /// <summary> Constructs a new <code>RenderObject</code>.</summary>
        public RenderObject()
        {
            baseTransform = new Matrix();
            transform = new Matrix();
        }

        /// <summary> Constructs a new <code>RenderObject</code> that is a copy of another
        /// <code>RenderObject</code>.
        /// </summary>
        public RenderObject(RenderObject ro)
        {
            baseTransform = ro.baseTransform.Clone();
            transform = ro.transform.Clone();
            frame = ro.frame;
        }

        /// <summary> 
        /// Paint this object.
        /// </summary>
        /// <param name="g">the graphics context where this object must be painted.</param>
        public abstract void Paint(Graphics g);

        /// <summary>
        /// Returns a copy of this object.
        /// </summary>
        public abstract object Clone();
    }
}